pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/BrokersBase.java

```diff
@@ -196,4 +210,90 @@ public InternalConfigurationData getInternalConfigurationData() {
+
+        CompletableFuture.allOf(producerFuture, readerFuture).whenComplete(
+                (ignore, exception) -> {
+                    if (exception != null) {
+                        completePromise.completeExceptionally(exception);
+                    } else {
+                        producerFuture.thenCompose((producer) -> producer.sendAsync(messageStr))
+                            .whenComplete((ignore2, exception2) -> {
+                                    if (exception2 != null) {
+                                        completePromise.completeExceptionally(exception2);
+                                    }
+                                });
+
+                        healthcheckReadLoop(readerFuture, completePromise, messageStr);
+                    }
+                });
+
+        completePromise.whenComplete((ignore, exception) -> {
+                producerFuture.thenAccept((producer) -> {
+                        producer.closeAsync().whenComplete((ignore2, exception2) -> {
+                                if (exception2 != null) {
+                                    LOG.warn("Error closing producer for healthcheck", exception2);
+                                }
+                            });
+                    });
+                readerFuture.thenAccept((reader) -> {
+                        reader.closeAsync().whenComplete((ignore2, exception2) -> {
+                                if (exception2 != null) {
+                                    LOG.warn("Error closing reader for healthcheck", exception2);
+                                }
+                            });
+                    });
+                if (exception != null) {
+                    asyncResponse.resume(new RestException(exception));
+                } else {
+                    asyncResponse.resume("ok");
+                }
+            });
+    }
+
+    private void healthcheckReadLoop(CompletableFuture<Reader<String>> readerFuture,
+                                     CompletableFuture<?> completablePromise,
+                                     String messageStr) {
+        readerFuture.thenAccept((reader) -> {
+                CompletableFuture<Message<String>> readFuture = reader.readNextAsync()
+                    .whenComplete((m, exception) -> {
+                            if (exception != null) {
+                                completablePromise.completeExceptionally(exception);
+                            } else if (m.getValue().equals(messageStr)) {
+                                completablePromise.complete(null);
+                            } else {
+                                healthcheckReadLoop(readerFuture, completablePromise, messageStr);
+                            }
+                        });
+                // timeout read after 10 seconds
+                ScheduledFuture<?> timeout = pulsar().getExecutor().schedule(() -> {
```

Comment:
isn't better just timeout at the outer level, so the timeout runnable will not be scheduled for each read.